![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2Fdata%2Bai%2Fbq-ai-functions%2Fworkflows%2Fmetric_diagnostics&file=metric_diagnostics.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2Fdata%252Bai%2Fbq-ai-functions%2Fworkflows%2Fmetric_diagnostics%2Fmetric_diagnostics.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ai-functions/workflows/metric_diagnostics/metric_diagnostics.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# Metric Diagnostics — BigQuery AI Functions

An end-to-end "why did my metric move?" pipeline:

1. **Build** a dataset with an interest period and a reference period
2. **Confirm** the headline shift with a plain SQL aggregate
3. **Explain** the change with `AI.KEY_DRIVERS` — find the segments that drove it
4. **Project** what each segment should have done with `AI.PREDICT` — a counterfactual built from reference-period behavior
5. **Test for cause** with `AI.CAUSAL_EFFECT` — and run the placebo check that any causal claim has to survive
6. **Narrate** the findings in plain language with `AI.GENERATE`

**What this demonstrates:**
- Root-cause analysis of a metric change entirely in BigQuery SQL
- Guarding a `SUM` metric against outlier records, and checking that dimension encodings are stable across the two periods, before trusting any decomposition
- Key driver / contribution analysis with `AI.KEY_DRIVERS` (no model, no connection)
- Ranking segments by absolute `contribution` and by `unexpected_difference`
- Zero-training tabular regression with `AI.PREDICT` (TabFM) as a per-segment counterfactual
- Two independent notions of "unexpected": statistical deviation and learned projection residual
- Composing augmented analytics with generative AI for an executive summary

**Functions used:** [`AI.KEY_DRIVERS`](../../functions/ai_key_drivers/) | [`AI.PREDICT`](../../functions/ai_predict/) | [`AI.CAUSAL_EFFECT`](../../../bq-ml/workflows/causal_effect/) | [`AI.GENERATE`](../../functions/ai_generate/)

**Prerequisites:** [Setup guide](../../setup/) | [Function reference](../../RESOURCES.md)

---
## Setup

Set your project and location, authenticate, and create a shared dataset.

> `AI.KEY_DRIVERS` and `AI.PREDICT` use end-user credentials and need no connection or model. `AI.GENERATE` routes to Gemini using a BigQuery connection — see the [Setup Reference](../../setup/) for details.

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
LOCATION = 'US'  # BigQuery dataset location
DATASET_ID = 'bq_ai_functions'  # Shared dataset across all notebooks
CONNECTION_ID = 'bq_ai_functions'  # Shared connection for AI.GENERATE

### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [Setup Reference](../../setup/) for details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [2]:
import subprocess, sys, shutil

def install(*packages):
    """Install packages using uv (fast) with pip fallback."""
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-cloud-bigquery', 'db-dtypes', 'bigquery-magics', 'tqdm')

In [3]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [4]:
from google.cloud import bigquery
import pandas as pd

client = bigquery.Client(project=PROJECT_ID)
pd.set_option('display.max_colwidth', None)

# Create the shared dataset (idempotent)
dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
dataset = bigquery.Dataset(dataset_ref)
dataset.location = LOCATION
client.create_dataset(dataset, exists_ok=True)
print(f'Dataset {PROJECT_ID}.{DATASET_ID} ready')

# Register %%bigquery cell magic (auto-loaded in Colab, needed elsewhere)
%load_ext bigquery_magics

Dataset statmike-mlops-349915.bq_ai_functions ready


### Connection for AI.GENERATE

`AI.GENERATE` (Step 6) needs a BigQuery Cloud resource connection with the Vertex AI User role. This is idempotent — skip if you already created it in another notebook.

In [5]:
import subprocess as _sp, json as _json

# Create connection (idempotent)
_sp.run(['bq', 'mk', '--connection', '--location', LOCATION,
         '--connection_type', 'CLOUD_RESOURCE',
         '--project_id', PROJECT_ID, CONNECTION_ID],
        capture_output=True, text=True)

# Get service account and grant the Vertex AI User role
r = _sp.run(['bq', 'show', '--connection', '--format=json',
             '--project_id', PROJECT_ID, '--location', LOCATION, CONNECTION_ID],
            capture_output=True, text=True, check=True)
sa = _json.loads(r.stdout)['cloudResource']['serviceAccountId']
_sp.run(['gcloud', 'projects', 'add-iam-policy-binding', PROJECT_ID,
         f'--member=serviceAccount:{sa}', '--role=roles/aiplatform.user', '--quiet'],
        capture_output=True, text=True)
print(f'Connection {CONNECTION_ID} ready (SA: {sa})')

Connection bq_ai_functions ready (SA: bqcx-1026793852137-hx0g@gcp-sa-bigquery-condel.iam.gserviceaccount.com)


---
## Step 1 — Build the dataset

We use the public [NYC Citi Bike](https://console.cloud.google.com/marketplace/details/city-of-new-york/nyc-citi-bike) trips dataset and compare two periods of the same season: **April 2017 (interest)** vs **April 2016 (reference)**. The metric is total trip duration; the dimensions are user type, gender, and start station.

**Guard the metric before you analyze it.** Citi Bike records a bike that is never returned as one enormous trip — the raw feed contains April trips lasting weeks. A `SUM` metric has no defense against that: a handful of such records can carry a double-digit share of the total and land at the top of every ranking downstream, where they read as changed rider behavior instead of missing hardware. The `tripduration BETWEEN 60 AND 86400` bound keeps trips from one minute to one day and drops the rest; the cell below prints how many records that removes and how much of the metric they were carrying, so the size of the problem is visible rather than assumed. Any sum-based diagnostic needs an equivalent bound chosen from the domain, not from the data.

Materializing a compact slice keeps every downstream query fast.

In [6]:
query = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_trips` AS
SELECT
  tripduration,
  usertype,
  gender,
  start_station_name,
  (EXTRACT(YEAR FROM starttime) = 2017) AS is_interest
FROM `bigquery-public-data.new_york_citibike.citibike_trips`
WHERE EXTRACT(MONTH FROM starttime) = 4
  AND EXTRACT(YEAR FROM starttime) IN (2016, 2017)
  AND tripduration IS NOT NULL
  -- Guard: 1 minute to 1 day. Unreturned bikes are logged as multi-week trips.
  AND tripduration BETWEEN 60 AND 86400
"""
client.query(query).result()

# What the duration guard removed
guard = client.query(f"""
  SELECT
    COUNTIF(NOT keep) AS trips_dropped,
    ROUND(SAFE_DIVIDE(COUNTIF(NOT keep), COUNT(*)), 6) AS share_of_trips_dropped,
    SUM(IF(keep, 0, tripduration)) AS seconds_dropped,
    ROUND(SAFE_DIVIDE(SUM(IF(keep, 0, tripduration)), SUM(tripduration)), 4) AS share_of_metric_dropped
  FROM (
    SELECT tripduration, tripduration BETWEEN 60 AND 86400 AS keep
    FROM `bigquery-public-data.new_york_citibike.citibike_trips`
    WHERE EXTRACT(MONTH FROM starttime) = 4
      AND EXTRACT(YEAR FROM starttime) IN (2016, 2017)
      AND tripduration IS NOT NULL
  )
""").to_dataframe().iloc[0]
print(f"Duration guard dropped {int(guard.trips_dropped):,} trips "
      f"({guard.share_of_trips_dropped:.4%} of records) "
      f"carrying {guard.share_of_metric_dropped:.2%} of the raw metric")

stats = client.query(f"""
  SELECT
    IF(is_interest, 'Interest (Apr 2017)', 'Reference (Apr 2016)') AS period,
    COUNT(*) AS trips,
    SUM(tripduration) AS total_duration
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_trips`
  GROUP BY is_interest
  ORDER BY is_interest
""").to_dataframe()
stats

Duration guard dropped 744 trips (0.0320% of records) carrying 14.22% of the raw metric


,period,trips,total_duration
0,Reference (Apr 2016),1012720,952691609
1,Interest (Apr 2017),1315089,1206124029


---
## Step 2 — Confirm the headline shift

Before asking *why*, confirm *what* changed. A plain SQL aggregate shows the overall move in total trip duration between the reference and interest periods. This is the number we want `AI.KEY_DRIVERS` to explain.

In [7]:
query = f"""
SELECT
  SUM(IF(NOT is_interest, tripduration, 0)) AS reference_total,
  SUM(IF(is_interest, tripduration, 0)) AS interest_total,
  SUM(IF(is_interest, tripduration, 0)) - SUM(IF(NOT is_interest, tripduration, 0)) AS difference,
  ROUND(SAFE_DIVIDE(
    SUM(IF(is_interest, tripduration, 0)) - SUM(IF(NOT is_interest, tripduration, 0)),
    SUM(IF(NOT is_interest, tripduration, 0))
  ), 3) AS relative_change
FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_trips`
"""
client.query(query).to_dataframe()

,reference_total,interest_total,difference,relative_change
0,952691609,1206124029,253432420,0.266


### Do the dimensions mean the same thing in both periods?

A driver analysis compares two periods through a set of dimension values, and it assumes those values are coded the same way on both sides. When a field's encoding changes between periods, segments appear and disappear, and the decomposition attributes the *bookkeeping* change to behavior. This check runs before the decomposition, not after, because it decides how the output can be read.

`interest_over_reference` is the tell. Overall trip volume grew by roughly a third, so a stable bucket lands near that. A `NULL` means the bucket had no reference-period trips at all, and a ratio far above the population's means riders were reclassified into it.

In [8]:
stability = client.query(f"""
  SELECT
    usertype,
    gender,
    COUNTIF(NOT is_interest) AS reference_trips,
    COUNTIF(is_interest) AS interest_trips,
    ROUND(SAFE_DIVIDE(COUNTIF(is_interest), COUNTIF(NOT is_interest)), 2) AS interest_over_reference
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_trips`
  GROUP BY usertype, gender
  ORDER BY usertype, gender
""").to_dataframe()

new_buckets = stability.loc[stability['reference_trips'] == 0, ['usertype', 'gender']]
print('usertype x gender buckets with zero reference-period trips '
      '(new in the interest period, not growth):')
print('  ' + ('\n  '.join(f'{u} / {g}' for u, g in new_buckets.itertuples(index=False))
              if len(new_buckets) else 'none'))
stability

usertype x gender buckets with zero reference-period trips (new in the interest period, not growth):
  Customer / female
  Customer / male


,usertype,gender,reference_trips,interest_trips,interest_over_reference
0,Customer,female,0,12954,NaN
1,Customer,male,0,17656,NaN
2,Customer,unknown,130249,158075,1.21
3,Subscriber,female,206442,274711,1.33
4,Subscriber,male,673988,837074,1.24
5,Subscriber,unknown,2041,14619,7.16


---
## Step 3 — Explain the change with AI.KEY_DRIVERS

`AI.KEY_DRIVERS` decomposes the headline shift into the segments that drove it. Each row is a segment (`drivers`), with its `contribution` (absolute size of the move) and `unexpected_difference` (how much it deviated from the overall trend).

One of those rows is not a segment: `AI.KEY_DRIVERS` also emits an `all` row carrying the population-level move. That row is a free cross-check — its `difference` should equal the `difference` Step 2 computed in plain SQL — and it is excluded from the segment rankings that follow.

We persist the result so Step 6 can summarize it.

In [9]:
query = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_drivers` AS
SELECT
  ARRAY_TO_STRING(drivers, ', ') AS segment,
  metric_interest,
  metric_reference,
  difference,
  relative_difference,
  unexpected_difference,
  apriori_support,
  contribution
FROM AI.KEY_DRIVERS(
  TABLE `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_trips`,
  metric_col => 'tripduration',
  dimension_cols => ['usertype', 'gender', 'start_station_name'],
  interest_label_col => 'is_interest',
  top_k => 20
)
"""
client.query(query).result()

# Top drivers by absolute contribution
client.query(f"""
  SELECT segment, metric_interest, metric_reference, difference, contribution
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_drivers`
  ORDER BY contribution DESC
  LIMIT 10
""").to_dataframe()

,segment,metric_interest,metric_reference,difference,contribution
0,all,1206124029,952691609,253432420,253432420
1,usertype=Subscriber,871764609,690392547,181372062,181372062
2,gender=male,650731415,507142971,143588444,143588444
3,"usertype=Subscriber, gender=male",621665152,507142971,114522181,114522181
4,gender=female,261923337,181585862,80337475,80337475
5,usertype=Customer,334359420,262299062,72060358,72060358
6,"usertype=Subscriber, gender=female",238430224,181585862,56844362,56844362
7,gender=unknown,293469277,263962776,29506501,29506501
8,"usertype=Customer, gender=male",29066263,0,29066263,29066263
9,"usertype=Customer, gender=female",23493113,0,23493113,23493113


### Segments defying the overall trend

The biggest absolute movers are often just the biggest segments. `unexpected_difference` highlights segments that changed *differently* than the population would predict — frequently the more actionable insight.

Read it next to `metric_reference`. A segment whose reference value is `0` did not exist in the reference period at all, so its entire interest-period metric registers as unexpected. Those rows rank on encoding, not behavior — which is exactly what the stability check above predicted for the `Customer` gender buckets.

In [10]:
client.query(f"""
  SELECT segment, metric_reference, difference, unexpected_difference, apriori_support
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_drivers`
  WHERE segment != 'all'   -- the population row, not a segment
  ORDER BY ABS(unexpected_difference) DESC
  LIMIT 10
""").to_dataframe()

,segment,metric_reference,difference,unexpected_difference,apriori_support
0,"usertype=Customer, gender=unknown",262299062,19500982,-6.937599e+07,0.275324
1,gender=unknown,263962776,29506501,-5.631553e+07,0.277071
2,"usertype=Subscriber, gender=male",507142971,114522181,-4.359154e+07,0.532326
3,gender=female,181585862,80337475,3.957576e+07,0.217161
4,"usertype=Customer, gender=male",0,29066263,2.906626e+07,0.024099
5,"usertype=Customer, gender=female",0,23493113,2.349311e+07,0.019478
6,gender=male,507142971,143588444,1.855922e+07,0.539523
7,"usertype=Subscriber, gender=female",181585862,56844362,1.055031e+07,0.197683
8,"usertype=Subscriber, gender=unknown",1663714,10005519,9.579672e+06,0.009675
9,usertype=Subscriber,690392547,181372062,-8.296684e+06,0.724676


---
## Step 4 — Project the metric per segment with AI.PREDICT

`AI.KEY_DRIVERS` explains the past: it decomposes a move that already happened. `AI.PREDICT` asks a different question — *what should the interest period have looked like if reference-period behavior had held?* The gap between that projection and the actual number is a second, independent kind of "unexpected".

`AI.PREDICT` runs the built-in TabFM tabular foundation model. There is no `CREATE MODEL`, no connection, and no training job: the training rows travel with the query as in-context examples. That design sets the ground rule for this step — **the training input has to stay small.** The observed ceiling is low and undocumented: roughly 10,000 training rows fails with `Resources exceeded ... allotted memory`, 8,000 succeeds, and about 5,000 is the safe budget on on-demand slots (see [`AI.PREDICT`](../../functions/ai_predict/)). The trip table holds millions of rows, so we first roll it up to one row per segment per period, keeping only segments that appear in both periods with enough volume to model. The cell below prints the resulting segment count — the training input is the reference-period half of it, and it needs to land comfortably under that budget.

In [11]:
query = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_segments` AS
WITH segments AS (
  SELECT
    usertype,
    gender,
    start_station_name,
    is_interest,
    COUNT(*) AS trips,
    SUM(tripduration) AS total_duration
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_trips`
  GROUP BY usertype, gender, start_station_name, is_interest
),
comparable AS (
  -- Only segments present in BOTH periods, with enough trips to be worth modeling
  SELECT usertype, gender, start_station_name
  FROM segments
  GROUP BY usertype, gender, start_station_name
  HAVING COUNT(DISTINCT is_interest) = 2 AND MIN(trips) >= 50
)
SELECT s.*
FROM segments AS s
JOIN comparable AS c
  ON s.usertype = c.usertype
 AND s.gender = c.gender
 AND s.start_station_name = c.start_station_name
"""
client.query(query).result()

client.query(f"""
  SELECT
    IF(is_interest, 'Interest (Apr 2017)', 'Reference (Apr 2016)') AS period,
    COUNT(*) AS segments,
    SUM(trips) AS trips,
    SUM(total_duration) AS total_duration
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_segments`
  GROUP BY is_interest
  ORDER BY is_interest
""").to_dataframe()

,period,segments,trips,total_duration
0,Reference (Apr 2016),1204,931567,877455091
1,Interest (Apr 2017),1204,1069183,942657220


### Project the interest period from reference-period behavior

The projection trains on the **reference** rows and predicts on the **interest** rows:

- **Features:** `usertype`, `gender`, `start_station_name`, and `trips` — four columns, well inside the documented 20-feature limit. `trips` is what makes this a counterfactual instead of an echo: the prediction row carries the *April 2017* ride volume, so the model answers "given this segment at this volume, what total duration would April 2016 behavior imply?" Without it, a segment's reference and interest rows would be identical and the projection would simply repeat last year's number.
- **Label:** `total_duration`. It is `INT64`, so `AI.PREDICT` runs a regression and returns `predicted_total_duration`. A `STRING` or `BOOL` label would trigger classification instead — the task is chosen by the label's type, not by an argument. That cuts both ways: a categorical encoded as `INT64` (a 0/1 flag, a 1–5 rating) is silently treated as regression and comes back as fractions with no probabilities, so cast those to `STRING` before predicting.
- **Types:** features must be `STRING`, `BOOL`, `INT64`, `FLOAT64`, `NUMERIC` or `BIGNUMERIC`. Step 1 already reduced `starttime` to a boolean period flag, which is why there is no timestamp column here to trip over.
- **Deliberately absent from the prediction input:** the actual `total_duration`. The prediction input is allowed to carry extra columns, but handing the model the answer it is predicting is a risk with no upside — the actuals are joined back afterward on the segment keys.
- **Which rows come back:** the returned passthrough columns are the *prediction* relation's rows, one per prediction row. The reference page says they come from the training table; the observed behavior — and Google's own worked examples — say otherwise, and this notebook depends on the behavior. It is why `p.trips` and the join keys in the queries below are April 2017 values, and why joining `p` to the interest-period segments is correct rather than backwards. See [`AI.PREDICT`](../../functions/ai_predict/) for the discrepancy.

`AI.PREDICT` is in Preview and takes tens of seconds per call regardless of input size. Materializing the result lets the ranking below and the narration in Step 6 both read it without paying for inference twice. Its output is also not reproducible run to run, which is a second reason to materialize rather than re-issue the call.

In [12]:
query = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_projection` AS
SELECT *
FROM AI.PREDICT(
  -- Training data: reference period (April 2016) behavior
  (SELECT usertype, gender, start_station_name, trips, total_duration
   FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_segments`
   WHERE NOT is_interest),
  -- Prediction data: the same segments at their interest period (April 2017) volumes
  (SELECT usertype, gender, start_station_name, trips
   FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_segments`
   WHERE is_interest),
  label_col => 'total_duration'
)
"""
client.query(query).result()

# Portfolio view: reference, projection, actual — plus a naive volume-scaled baseline
portfolio = client.query(f"""
  SELECT
    COUNT(*) AS segments,
    SUM(r.total_duration) AS reference_total,
    CAST(ROUND(SUM(p.predicted_total_duration)) AS INT64) AS projected_total,
    SUM(a.total_duration) AS actual_total,
    SUM(CAST(ROUND(a.trips * SAFE_DIVIDE(r.total_duration, r.trips)) AS INT64)) AS volume_scaled_total,
    COUNTIF(a.total_duration < p.predicted_total_duration) AS segments_below_projection,
    CAST(ROUND(SUM(GREATEST(a.total_duration - p.predicted_total_duration, 0))) AS INT64) AS total_overshoot,
    CAST(ROUND(SUM(LEAST(a.total_duration - p.predicted_total_duration, 0))) AS INT64) AS total_shortfall
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_projection` AS p
  JOIN `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_segments` AS a
    ON a.is_interest
   AND a.usertype = p.usertype
   AND a.gender = p.gender
   AND a.start_station_name = p.start_station_name
  JOIN `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_segments` AS r
    ON NOT r.is_interest
   AND r.usertype = p.usertype
   AND r.gender = p.gender
   AND r.start_station_name = p.start_station_name
""").to_dataframe()

f = portfolio.iloc[0]
print(f"Projection tracks volume, not last year: projected {f.projected_total:,} vs "
      f"volume-scaled {f.volume_scaled_total:,} vs reference {f.reference_total:,}")
print(f"Segments below projection: {f.segments_below_projection:,} of {f.segments:,} "
      f"({f.segments_below_projection / f.segments:.0%})")
print(f"Aggregate gap: actual - projected = {f.actual_total - f.projected_total:,} s "
      f"({(f.actual_total - f.projected_total) / f.projected_total:+.1%})")
print(f"  broad shortfall across the below-projection segments: {f.total_shortfall:,} s")
print(f"  concentrated overshoot across the above-projection segments: +{f.total_overshoot:,} s")
portfolio

Projection tracks volume, not last year: projected 995,052,044 vs volume-scaled 999,847,409 vs reference 877,455,091
Segments below projection: 839 of 1,204 (70%)
Aggregate gap: actual - projected = -52,394,824 s (-5.3%)
  broad shortfall across the below-projection segments: -91,696,549 s
  concentrated overshoot across the above-projection segments: +39,301,725 s


,segments,reference_total,projected_total,actual_total,volume_scaled_total,segments_below_projection,total_overshoot,total_shortfall
0,1204,877455091,995052044,942657220,999847409,839,39301725,-91696549


`reference_total` is what these segments recorded in April 2016, `projected_total` is what `AI.PREDICT` expects from April 2017 volumes under April 2016 behavior, and `actual_total` is what April 2017 actually recorded. `volume_scaled_total` is the naive alternative — each segment's reference seconds-per-trip multiplied by its interest-period trip count — and it is the sanity check on the whole step: a `projected_total` that lands on `reference_total` rather than tracking volume means the model is repeating last year instead of projecting, and the residuals below would carry no information.

`segments_below_projection` is the shape of the result, and it deserves reading rather than skipping. Most of the segments come in *under* their projection, and the portfolio as a whole falls a few percent short — riders took more trips in April 2017 but each trip was a little shorter, so scaling last year's seconds-per-trip by this year's volume overshoots almost everywhere. The printed split makes the two pictures reconcile: a broad negative shortfall spread over the majority of segments, against a smaller positive overshoot concentrated in a few. The ranking below is ordered by *absolute* residual precisely so both tails appear in it, instead of showing only the segments that ran long.

### Segments that missed their projection

The residual is `actual_total - projected_total`: negative where a segment delivered less duration than its volume implied, positive where it delivered more. Ranking on `ABS(residual)` shows both directions.

`max_trip_share` is the guard from Step 1 earning its keep — the longest single trip in the segment as a fraction of the segment's actual total. On a sum-based metric this column is the difference between a finding and an artifact: if it is a few percent, the residual describes many riders, and if it approaches 1 the residual is one record. Carry a column like it into any residual table you build.

In [13]:
client.query(f"""
  WITH peaks AS (
    -- Longest single trip per interest-period segment: the artifact detector
    SELECT usertype, gender, start_station_name, MAX(tripduration) AS max_trip
    FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_trips`
    WHERE is_interest
    GROUP BY usertype, gender, start_station_name
  )
  SELECT
    p.usertype,
    p.gender,
    p.start_station_name,
    r.trips AS reference_trips,
    p.trips AS interest_trips,
    r.total_duration AS reference_total,
    CAST(ROUND(p.predicted_total_duration) AS INT64) AS projected_total,
    a.total_duration AS actual_total,
    CAST(ROUND(a.total_duration - p.predicted_total_duration) AS INT64) AS residual,
    ROUND(SAFE_DIVIDE(k.max_trip, a.total_duration), 4) AS max_trip_share
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_projection` AS p
  JOIN `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_segments` AS a
    ON a.is_interest
   AND a.usertype = p.usertype
   AND a.gender = p.gender
   AND a.start_station_name = p.start_station_name
  JOIN `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_segments` AS r
    ON NOT r.is_interest
   AND r.usertype = p.usertype
   AND r.gender = p.gender
   AND r.start_station_name = p.start_station_name
  JOIN peaks AS k
    ON k.usertype = p.usertype
   AND k.gender = p.gender
   AND k.start_station_name = p.start_station_name
  ORDER BY ABS(a.total_duration - p.predicted_total_duration) DESC
  LIMIT 10
""").to_dataframe()

,usertype,gender,start_station_name,reference_trips,interest_trips,reference_total,projected_total,actual_total,residual,max_trip_share
0,Customer,unknown,12 Ave & W 40 St,2103,2544,3966788,6160380,4365733,-1794647,0.0162
1,Customer,unknown,Centre St & Chambers St,2093,3071,4520343,7471100,5932719,-1538381,0.0135
2,Subscriber,male,12 Ave & W 40 St,2837,3137,3126439,2228220,3683543,1455323,0.0157
3,Subscriber,male,West St & Chambers St,3793,4688,3720234,3309570,4460041,1150471,0.0059
4,Customer,unknown,Central Park S & 6 Ave,3722,4635,10275969,10551300,9433040,-1118260,0.0086
5,Subscriber,male,Pershing Square North,9605,10067,7379797,6389760,7416847,1027087,0.0033
6,Customer,unknown,5 Ave & E 73 St,1784,2523,4319049,5865470,4903091,-962379,0.0176
7,Customer,unknown,West St & Chambers St,2076,1984,4008799,4177920,3227137,-950783,0.0188
8,Subscriber,male,University Pl & E 14 St,3584,6072,2294784,4718590,3787487,-931103,0.0114
9,Subscriber,male,W 21 St & 6 Ave,5550,6166,3504499,4784130,3858605,-925525,0.0097


`residual` is a *learned* expectation error: actual minus what a model of reference-period behavior expects at interest-period volume. `unexpected_difference` from Step 3 is a *statistical* expectation error: how far a segment moved from what the population-wide trend implied for it.

The two rankings disagree on purpose. A segment can be unremarkable to `AI.KEY_DRIVERS` — it moved with everyone else — and still miss its projection badly because its riders changed how long they ride. Reading both separates "this segment is large" from "this segment behaves differently now", and Step 6 hands both to the narrator.

Before drawing that behavioral reading, though, check `max_trip_share` on every row you plan to act on. Only a residual backed by a small `max_trip_share` is a statement about riders; a large one is a statement about a single record, and no amount of modeling downstream will fix it.

---
## Step 5 — Was anything *caused*? `AI.CAUSAL_EFFECT` and the placebo test

Everything so far describes the move. `AI.KEY_DRIVERS` said **where** it concentrated; `AI.PREDICT` said **which segments missed a projection built from their own past**. Neither is a causal statement, and neither becomes one by being run on before-and-after periods.

`AI.CAUSAL_EFFECT` is the function that asks the causal question, and it wants a different shape of data: **one series, a time axis, and a dated intervention**. It fits an `ARIMA_PLUS` counterfactual on the pre-intervention stretch, projects it forward, and reports the cumulative gap with a p-value. It takes no control group and no covariates, so its entire identifying assumption is *the series would have continued its own pattern*. The mechanics are measured end to end in the sibling project's [`bq-ml/workflows/causal_effect/`](../../../bq-ml/workflows/causal_effect/) — including the finding that its counterfactual is reproducible bit-for-bit with a plain `CREATE MODEL ... ARIMA_PLUS`.

**There is no known intervention in this dataset**, which makes it the right place to demonstrate the check that belongs next to every causal claim rather than the claim itself. A **placebo test** runs the same analysis at dates where nothing happened. If the function reports an effect there too, the design cannot distinguish an intervention from the series' own behaviour, and no p-value from it means what it appears to mean.

One data note first: `citibike_trips` has a gap from October 2016 through March 2017, so a continuous daily series cannot span both comparison periods. The placebo test runs on the continuous stretch that does exist, **2017-04-01 through 2018-05-31**.

In [14]:
query = f"""
CREATE OR REPLACE TABLE `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_daily` AS
SELECT
  TIMESTAMP(DATE(starttime)) AS day_ts,
  SUM(tripduration) / 3600.0 AS total_hours
FROM `bigquery-public-data.new_york_citibike.citibike_trips`
WHERE DATE(starttime) BETWEEN '2017-04-01' AND '2018-05-31'
  AND tripduration IS NOT NULL
  -- the same outlier guard as Step 1, for the same reason
  AND tripduration BETWEEN 60 AND 86400
GROUP BY day_ts
"""
client.query(query).result()

client.query(f"""
  SELECT
    COUNT(*) AS days,
    MIN(day_ts) AS first_day,
    MAX(day_ts) AS last_day,
    ROUND(AVG(total_hours)) AS avg_daily_hours
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_daily`
""").to_dataframe()

,days,first_day,last_day,avg_daily_hours
0,426,2017-04-01 00:00:00+00:00,2018-05-31 00:00:00+00:00,11115.0


In [15]:
# Six dates chosen only for spacing across the series. None is a real intervention.
# num_post_intervention_points holds the post window at 60 days so the six tests are comparable.
no_cache = bigquery.QueryJobConfig(use_query_cache=False)

def placebo(intervention, post_points=60):
    return client.query(f"""
    SELECT absolute_effect, relative_effect, p_value
    FROM AI.CAUSAL_EFFECT(
      (SELECT day_ts, total_hours
       FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_daily`),
      timestamp_col                => 'day_ts',
      data_col                     => 'total_hours',
      intervention_timestamp       => TIMESTAMP '{intervention}',
      num_post_intervention_points => {post_points})
    """, job_config=no_cache).to_dataframe().iloc[0]

dates = ['2017-07-01', '2017-09-01', '2017-10-01', '2017-12-01', '2018-01-01', '2018-03-01']
tests = pd.DataFrame({d: placebo(d) for d in dates}).T
tests.index.name = 'placebo intervention date'
tests['significant at 0.05'] = tests['p_value'] < 0.05
tests

,absolute_effect,relative_effect,p_value,significant at 0.05
placebo intervention date,,,,
2017-07-01,-135040.470288,-0.131746,0.521827,False
2017-09-01,-117060.749421,-0.114272,0.425137,False
2017-10-01,-302250.389426,-0.293742,0.044108,True
2017-12-01,-255041.113855,-0.457276,0.341785,False
2018-01-01,147215.092995,0.882366,0.602125,False
2018-03-01,41044.408046,0.089598,0.879161,False


**Verified finding — one placebo date out of six clears `p < 0.05`, and it is the one the seasonality predicts.** Nothing happened on any of these dates. Five return effects the function itself declines to call significant. **2017-10-01 does not** — it reports a large negative cumulative effect at a p-value under 0.05, because a counterfactual fitted on a spring-and-summer ramp cannot know that ridership falls every autumn, so it carries the summer level forward and reads the seasonal decline as an intervention.

That is the whole lesson in one row. The function is behaving exactly as documented; the failure is in the design around it. Nor is the result a borderline artifact of a liberal test — the sibling project shows this p-value is computed on the perfect-positive-correlation *ceiling* variance, the most conservative construction available, so `0.044108` is the **largest** p-value that reading of the data supports. A univariate counterfactual can only carry forward patterns it has already seen, and an annual cycle needs more than a year of pre-intervention data before it is one of them. Point the same call at a date just before a seasonal turn, with a pre-window shorter than the cycle, and it will hand you a significant result with no cause behind it.

**What this means for the diagnostic above.** `AI.KEY_DRIVERS` and `AI.PREDICT` earned their findings without claiming cause, and that restraint is not a limitation to be fixed by adding a causal function — it is the honest reading of two-period data with no intervention in it. When a real dated intervention *does* exist, `AI.CAUSAL_EFFECT` is the right tool, and the discipline that makes its answer usable is this table: run the placebo dates first, and treat the real date's p-value as informative only if the placebos stay quiet.

Two other defences are worth knowing, both demonstrated in the sibling project: give the pre-window enough history to contain a full seasonal cycle, and where a credible control unit exists, prefer a method that uses one — [`difference_in_differences`](../../../bq-ml/workflows/difference_in_differences/) or [`synthetic_control`](../../../bq-ml/workflows/synthetic_control/), which net out shocks a univariate counterfactual has no way to see.

---
## Step 6 — Narrate the findings with AI.GENERATE

`AI.KEY_DRIVERS` gives us the contributions and `AI.PREDICT` gives us the projection residuals; `AI.GENERATE` turns both into a plain-language executive summary. We aggregate the top driver rows and the largest residuals into a single prompt and ask Gemini to explain what drove the metric change.

In [16]:
query = f"""
WITH drivers AS (
  SELECT STRING_AGG(
    CONCAT(
      segment,
      ': difference ', CAST(ROUND(difference) AS STRING),
      ', unexpected ', CAST(ROUND(unexpected_difference) AS STRING),
      ', reference ', CAST(ROUND(metric_reference) AS STRING),
      ', support ', CAST(ROUND(apriori_support, 3) AS STRING)
    ),
    ' ||| '
    ORDER BY contribution DESC
    LIMIT 12
  ) AS driver_rows
  FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_drivers`
  WHERE segment != 'all'
),
projections AS (
  SELECT STRING_AGG(
    CONCAT(
      usertype, ' / ', gender, ' / ', start_station_name,
      ': projected ', CAST(projected_total AS STRING),
      ', actual ', CAST(actual_total AS STRING),
      ', residual ', CAST(residual AS STRING)
    ),
    ' ||| '
    ORDER BY ABS(residual) DESC
    LIMIT 8
  ) AS projection_rows
  FROM (
    SELECT
      p.usertype,
      p.gender,
      p.start_station_name,
      CAST(ROUND(p.predicted_total_duration) AS INT64) AS projected_total,
      a.total_duration AS actual_total,
      CAST(ROUND(a.total_duration - p.predicted_total_duration) AS INT64) AS residual
    FROM `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_projection` AS p
    JOIN `{PROJECT_ID}.{DATASET_ID}.workflow_metricdiag_segments` AS a
      ON a.is_interest
     AND a.usertype = p.usertype
     AND a.gender = p.gender
     AND a.start_station_name = p.start_station_name
  )
)
SELECT (AI.GENERATE(
  CONCAT(
    'You are a data analyst. Total NYC Citi Bike trip duration changed between April 2016 (reference) and ',
    'April 2017 (interest). The following segments were identified as key drivers of that change by a ',
    'contribution analysis. Each entry lists the segment, its absolute difference in trip-duration seconds, ',
    'its unexpected difference (deviation from the overall trend), its reference-period metric value, and ',
    'its apriori support (segment size). ',
    'A second list gives per-segment projections: the total trip duration the April 2017 ride volume would ',
    'have produced under April 2016 behavior, the actual April 2017 total, and the residual (actual minus ',
    'projected). ',
    'Write a concise executive summary: (1) the overall direction of the change, (2) which segments drove it most, ',
    '(3) any segments that moved differently than expected, (4) where the actual result diverged most from its ',
    'projection. Use plain business language. ',
    'Grounding rules: quote the supplied figures verbatim whenever you name a segment, and do not state any ',
    'ratio, multiple or percentage that is not given below. A segment whose reference value is 0 did not exist ',
    'in the reference period, so describe it as newly coded rather than as growth. ',
    'Key drivers: ', driver_rows,
    ' ||| Projection residuals: ', projection_rows
  )
)).result AS summary
FROM drivers, projections
"""
df = client.query(query).to_dataframe()
print(df.iloc[0]['summary'])

Here's an executive summary of the change in total NYC Citi Bike trip duration between April 2016 and April 2017:

Total trip duration in April 2017 increased compared to April 2016.

The segments that drove this increase most significantly were:
*   **usertype=Subscriber**: with a difference of 181372062 seconds.
*   **gender=male**: with a difference of 143588444 seconds.
*   **usertype=Subscriber, gender=male**: with a difference of 114522181 seconds.
*   **gender=female**: with a difference of 80337475 seconds.
*   **usertype=Customer**: with a difference of 72060358 seconds.

Several segments moved differently than expected based on the overall trend:
*   The **usertype=Subscriber** segment showed an unexpected decrease of -8296684 seconds.
*   The **usertype=Subscriber, gender=male** segment showed a notable unexpected decrease of -43591536 seconds.
*   The **gender=unknown** segment showed a significant unexpected decrease of -56315530 seconds.
*   The **usertype=Customer, gende

The prompt hands the model the driver rows and the residual rows as text and asks for prose, so the grounding rules in it are doing real work: quote the supplied figures, invent no ratios that were not given, and describe a zero reference value as a new encoding rather than as growth. Without those rules a summary of this shape reliably produces a plausible multiple or percentage that no input row supports.

Treat the result accordingly — it is a narrative layer, not a source of numbers. Every figure a stakeholder acts on should be joined back from `workflow_metricdiag_drivers` and `workflow_metricdiag_projection`, where it can be recomputed. That is the augmented-analytics division of labor: statistical functions find *what* drove the change and *what was expected instead*, and a generative function turns the rows already computed into language a stakeholder can act on.

---
## Cleanup

Drop tables created by this workflow. Shared resources (dataset, connection) are left for other notebooks.

In [17]:
for table in ['workflow_metricdiag_projection', 'workflow_metricdiag_segments',
              'workflow_metricdiag_drivers', 'workflow_metricdiag_daily',
              'workflow_metricdiag_trips']:
    client.delete_table(f'{PROJECT_ID}.{DATASET_ID}.{table}', not_found_ok=True)
    print(f'Dropped {table}')

Dropped workflow_metricdiag_projection


Dropped workflow_metricdiag_segments
Dropped workflow_metricdiag_drivers


Dropped workflow_metricdiag_daily
Dropped workflow_metricdiag_trips


### Remove all shared resources

When finished with **all** notebooks, uncomment and run the cell below to delete the shared dataset and all tables, models, and other resources within it.

In [18]:
# client.delete_dataset(
#     f'{PROJECT_ID}.{DATASET_ID}',
#     delete_contents=True,
#     not_found_ok=True
# )
# print(f'Dataset {PROJECT_ID}.{DATASET_ID} deleted')